### ISSUE TITLE
1Lab

### ISSUE THREAD

**Issue by dlcor** [dlcor/AI#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.preprocessing import StandardScaler

# Шаг 1: Загрузка и предварительная обработка данных
try:
    df = pd.read_csv("GeneralEsportData.csv")

    # Преобразование числовых колонок (удаление символов и преобразование в float)
    numeric_cols = ['TotalEarnings', 'OfflineEarnings', 'TotalPlayers', 'TotalTournaments']
    for col in numeric_cols:
        df[col] = pd.to_numeric(df[col].astype(str).str.replace(',', ''), errors='coerce')

    # Заполнение пропусков
    df['PercentOffline'] = df['PercentOffline'].fillna(0)
    df['OfflineEarnings'] = df['OfflineEarnings'].fillna(0)

    # Создание новых признаков
    df['EarningsPerPlayer'] = df['TotalEarnings'] / df['TotalPlayers'].replace(0, 1)
    df['EarningsPerTournament'] = df['TotalEarnings'] / df['TotalTournaments'].replace(0, 1)

    # Логарифмирование для нормализации распределения
    df['log_TotalEarnings'] = np.log1p(df['TotalEarnings'])

except Exception as e:
    print(f"Ошибка при загрузке данных: {e}")
    exit()

# Шаг 2: Анализ данных
print("\nПервые 5 строк данных:")
print(df.head())

print("\nИнформация о данных:")
print(df.info())

print("\nОписательная статистика:")
print(df.describe())

# Шаг 3: Визуализация данных
plt.figure(figsize=(15, 10))

# Распределение доходов по жанрам
plt.subplot(2, 2, 1)
sns.boxplot(x='Genre', y='TotalEarnings', data=df)
plt.xticks(rotation=45)
plt.title('Распределение доходов по жанрам')
plt.yscale('log')

# Топ 10 игр по общему доходу
plt.subplot(2, 2, 2)
top_games = df.sort_values('TotalEarnings', ascending=False).head(10)
sns.barplot(x='TotalEarnings', y='Game', data=top_games, palette='viridis')
plt.title('Топ 10 игр по общему доходу')

# Корреляционная матрица
plt.subplot(2, 2, 3)
numeric_df = df.select_dtypes(include=[np.number])
corr = numeric_df.corr()
sns.heatmap(corr, annot=True, cmap='coolwarm', fmt=".2f")
plt.title('Корреляционная матрица')

# Распределение доходов
plt.subplot(2, 2, 4)
sns.histplot(df['log_TotalEarnings'], bins=30, kde=True)
plt.title('Логарифм общего дохода (распределение)')

plt.tight_layout()
plt.show()

# Шаг 4: Подготовка данных для моделирования
# Фильтрация данных (удаление строк с нулевыми значениями в целевой переменной)
df = df[df['TotalEarnings'] > 0]

# Выбор признаков и целевой переменной
features = ['OfflineEarnings', 'TotalPlayers', 'TotalTournaments', 'PercentOffline']
X = df[features]
y = df['log_TotalEarnings']  # Используем логарифмированные значения

# Преобразование категориальных переменных (Genre)
genre_dummies = pd.get_dummies(df['Genre'], prefix='Genre', drop_first=True)
X = pd.concat([X, genre_dummies], axis=1)

# Масштабирование числовых признаков
scaler = StandardScaler()
X[features] = scaler.fit_transform(X[features])

# Разделение на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Шаг 5: Обучение модели линейной регрессии
model = LinearRegression()
model.fit(X_train, y_train)

# Шаг 6: Оценка модели
# Прогнозирование на тестовых данных
y_pred = model.predict(X_test)

# Обратное преобразование из логарифма
y_test_exp = np.expm1(y_test)
y_pred_exp = np.expm1(y_pred)

# Метрики качества
mae = mean_absolute_error(y_test_exp, y_pred_exp)
mse = mean_squared_error(y_test_exp, y_pred_exp)
rmse = np.sqrt(mse)

print("\nОценка модели:")
print(f"MAE: {mae:.2f}")
print(f"MSE: {mse:.2f}")
print(f"RMSE: {rmse:.2f}")
print(f"R^2 на обучающих данных: {model.score(X_train, y_train):.2f}")
print(f"R^2 на тестовых данных: {model.score(X_test, y_test):.2f}")

# Шаг 7: Визуализация результатов
plt.figure(figsize=(12, 5))

# График реальных vs предсказанных значений
plt.subplot(1, 2, 1)
sns.scatterplot(x=y_test_exp, y=y_pred_exp)
plt.plot([y_test_exp.min(), y_test_exp.max()], [y_test_exp.min(), y_test_exp.max()], 'r--')
plt.xlabel('Реальные значения (TotalEarnings)')
plt.ylabel('Предсказанные значения')
plt.title('Реальные vs Предсказанные значения')

# Важность признаков
coefficients = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': model.coef_
}).sort_values('Coefficient', ascending=False)

plt.subplot(1, 2, 2)
sns.barplot(x='Coefficient', y='Feature', data=coefficients)
plt.title('Важность признаков в модели')
plt.tight_layout()
plt.show()

# Шаг 8: Анализ остатков
residuals = y_test_exp - y_pred_exp
plt.figure(figsize=(10, 5))
sns.histplot(residuals, bins=30, kde=True)
plt.xlabel('Остатки (Реальные - Предсказанные)')
plt.title('Распределение остатков')
plt.show()
